[SPARK-7326] [STREAMING] Performing window() on a WindowedDStream doesn't work all the time

wesleymiao opened the pull request:
@tdas 

https://issues.apache.org/jira/browse/SPARK-7326

The problem most likely resides in DStream.slice() implementation, as shown below.

  def slice(fromTime: Time, toTime: Time): Seq[RDD[T]] = {
    if (!isInitialized) {
      throw new SparkException(this + " has not been initialized")
    }
    if (!(fromTime - zeroTime).isMultipleOf(slideDuration)) {
      logWarning("fromTime (" + fromTime + ") is not a multiple of slideDuration ("
        + slideDuration + ")")
    }
    if (!(toTime - zeroTime).isMultipleOf(slideDuration)) {
      logWarning("toTime (" + fromTime + ") is not a multiple of slideDuration ("
        + slideDuration + ")")
    }
    val alignedToTime = toTime.floor(slideDuration, zeroTime)
    val alignedFromTime = fromTime.floor(slideDuration, zeroTime)

```
logInfo("Slicing from " + fromTime + " to " + toTime +
  " (aligned to " + alignedFromTime + " and " + alignedToTime + ")")

alignedFromTime.to(alignedToTime, slideDuration).flatMap(time => {
  if (time >= zeroTime) getOrCompute(time) else None
})
```

  }

Here after performing floor() on both fromTime and toTime, the result (alignedFromTime - zeroTime) and (alignedToTime - zeroTime) may no longer be multiple of the slidingDuration, thus making isTimeValid() check failed for all the remaining computation.

The fix is to add a new floor() function in Time.scala to respect the zeroTime while performing the floor :

  def floor(that: Duration, zeroTime: Time): Time = {
    val t = that.milliseconds
    new Time(((this.millis - zeroTime.milliseconds) / t) \* t + zeroTime.milliseconds)
  }

And then change the DStream.slice to call this new floor function by passing in its zeroTime.

```
val alignedToTime = toTime.floor(slideDuration, zeroTime)
val alignedFromTime = fromTime.floor(slideDuration, zeroTime)
```

This way the alignedToTime and alignedFromTime are _really_ aligned in respect to zeroTime whose value is not really a 0.

wesleymiao committed:
[SPARK-7326] [STREAMING] Performing window() on a WindowedDStream doesn't work all the time

The problem most likely resides in DStream.slice() implementation, as shown below.

  def slice(fromTime: Time, toTime: Time): Seq[RDD[T]] = {
    if (!isInitialized) {
      throw new SparkException(this + " has not been initialized")
    }
    if (!(fromTime - zeroTime).isMultipleOf(slideDuration)) {
      logWarning("fromTime (" + fromTime + ") is not a multiple of slideDuration ("
        + slideDuration + ")")
    }
    if (!(toTime - zeroTime).isMultipleOf(slideDuration)) {
      logWarning("toTime (" + fromTime + ") is not a multiple of slideDuration ("
        + slideDuration + ")")
    }
    val alignedToTime = toTime.floor(slideDuration, zeroTime)
    val alignedFromTime = fromTime.floor(slideDuration, zeroTime)

    logInfo("Slicing from " + fromTime + " to " + toTime +
      " (aligned to " + alignedFromTime + " and " + alignedToTime + ")")

    alignedFromTime.to(alignedToTime, slideDuration).flatMap(time => {
      if (time >= zeroTime) getOrCompute(time) else None
    })
  }

Here after performing floor() on both fromTime and toTime, the result (alignedFromTime - zeroTime) and (alignedToTime - zeroTime) may no longer be multiple of the slidingDuration, thus making isTimeValid() check failed for all the remaining computation.

The fix is to add a new floor() function in Time.scala to respect the zeroTime while performing the floor :

  def floor(that: Duration, zeroTime: Time): Time = {
    val t = that.milliseconds
    new Time(((this.millis - zeroTime.milliseconds) / t) * t + zeroTime.milliseconds)
  }

And then change the DStream.slice to call this new floor function by passing in its zeroTime.

    val alignedToTime = toTime.floor(slideDuration, zeroTime)
    val alignedFromTime = fromTime.floor(slideDuration, zeroTime)

This way the alignedToTime and alignedFromTime are *really* aligned in respect to zeroTime whose value is not really a 0.

AmplabJenkins commented:
Can one of the admins verify this patch?

srowen commented:
The intent of the code does seem to be that the difference from `zeroTime` is what is supposed to be a multiple of `slideDuration`. From my understanding I agree with your change. Let's see what the tests say.

srowen commented:
ok to test

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #31691 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/31691/consoleFull) for   PR 5871 at commit [`2611745`](https://github.com/apache/spark/commit/261174533048be2823e01a82071648794a86da8d).

SparkQA commented:
[Test build #31691 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/31691/consoleFull) for   PR 5871 at commit [`2611745`](https://github.com/apache/spark/commit/261174533048be2823e01a82071648794a86da8d).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

viirya commented:
You will not directly call `slice()`. When `slice()` is called internally, e.g. in `WindowedDStream.compute()`, the `Interval` given as parameter to `slice` is valid.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/31691/
Test PASSed.

wesleymiao committed:
[SPARK-7326] [STREAMING] Performing window() on a WindowedDStream doesn't work all the time

Improve it a bit to only do floor when it is not aligned relative to its zeroTime

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #31715 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/31715/consoleFull) for   PR 5871 at commit [`6ade399`](https://github.com/apache/spark/commit/6ade399a3e7ba70391d1d089fe930e83c10ce6bd).

wesleymiao commented:
@viirya , you are right. I've changed a little bit to only perform floor when the fromTime and toTime are not aligned relative to zeroTime.

Also fixed a typo in the log message.

SparkQA commented:
[Test build #31715 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/31715/consoleFull) for   PR 5871 at commit [`6ade399`](https://github.com/apache/spark/commit/6ade399a3e7ba70391d1d089fe930e83c10ce6bd).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/31715/
Test PASSed.

srowen commented on streaming/src/main/scala/org/apache/spark/streaming/dstream/DStream.scala:
```diff
@@ -762,16 +762,22 @@ abstract class DStream[T: ClassTag] (
-      logWarning("fromTime (" + fromTime + ") is not a multiple of slideDuration ("
-        + slideDuration + ")")
+
+    val alignedToTime = (toTime - zeroTime).isMultipleOf(slideDuration) match {
```
I think match syntax for booleans is needlessly complicated. You can just write a more straightforward `if`.

wesleymiao replied:
changed to if-else.

wesleymiao committed:
[SPARK-7326] [STREAMING] Performing window() on a WindowedDStream doesn't work all the time

In response to comment, change match-case to if-else for boolean to be consistent.

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #31730 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/31730/consoleFull) for   PR 5871 at commit [`48b4dc0`](https://github.com/apache/spark/commit/48b4dc0effa141b9fd2e307accdc926b4fe64663).

SparkQA commented:
[Test build #31730 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/31730/consoleFull) for   PR 5871 at commit [`48b4dc0`](https://github.com/apache/spark/commit/48b4dc0effa141b9fd2e307accdc926b4fe64663).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/31730/
Test PASSed.

zsxwing commented on streaming/src/main/scala/org/apache/spark/streaming/Time.scala:
```diff
@@ -63,6 +63,11 @@ case class Time(private val millis: Long) {
     new Time((this.millis / t) * t)
   }
 
+  def floor(that: Duration, zeroTime: Time): Time = {
```
Could you also add some tests for this new method?

wesleymiao replied:
Added a few tests.

tdas replied:
Should this be made public? If so, would be good to add some docs.

srowen commented:
A little test would be great, yeah. Would still love a quick look from @tdas to make sure we're not missing something.

wesleymiao committed:
[SPARK-7326] [STREAMING] Performing window() on a WindowedDStream dosen't work all the time

Add some tests to cover the new function Time.floor(duration, zeroTime)

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #32230 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/32230/consoleFull) for   PR 5871 at commit [`82a4d8c`](https://github.com/apache/spark/commit/82a4d8c443ebc17e6aa5e4ed11672686adc03700).

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/32230/
Test FAILed.

shaneknapp commented:
i'll retrigger this job once i restart jenkins

shaneknapp commented:
jenkins, test this please

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #32237 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/32237/consoleFull) for   PR 5871 at commit [`82a4d8c`](https://github.com/apache/spark/commit/82a4d8c443ebc17e6aa5e4ed11672686adc03700).

SparkQA commented:
[Test build #32237 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/32237/consoleFull) for   PR 5871 at commit [`82a4d8c`](https://github.com/apache/spark/commit/82a4d8c443ebc17e6aa5e4ed11672686adc03700).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/32237/
Test PASSed.

wesleymiao commented:
Hi, @tdas , will you be able to take a look at this PR? My team in Autodesk is trying to adopt Spark Streaming to do some real-time analytics for our services' API logs to serve various purposes, among which is to monitor our services' health in real-time based on API log statistics. Then we run into this situation where we need to perform reduceByKeyAndWindow on an already windowed DStream. 

What I'd like to achieve is to do multiple-level reduce on the source of API logging stream. For example - let's say the source logging stream is at interval 1 second. The first level aggregation would be every 1 minute, which is the 60 intervals of the source stream. The second level would be every 1 hour. The third level would be 1 day. And we can do more levels if we want.

What I hope is that at each level we'll do reduceByKeyAndWindow so that its aggregation can be done over its immediate previous level, instead of always aggregating over the source stream. Level 3's reduceByKey will be based on level 2's result, level 2 is based one level 1 and level 1 is based on the source stream.

I would think this approach will be more efficient than always reducing over the source stream, particularly for the higher level (like daily and weekly) aggregation.

srowen commented:
I'm proceeding with this since everywhere in the code it appears that the intent is for RDDs to start an end on a multiple of the slide duration _after zero time_. Therefore it seems like the intent must have been to take into account `zeroTime`.

tdas commented on streaming/src/main/scala/org/apache/spark/streaming/dstream/DStream.scala:
```diff
@@ -762,16 +762,22 @@ abstract class DStream[T: ClassTag] (
+      fromTime
+    } else {
+      logWarning("fromTime (" + fromTime + ") is not a multiple of slideDuration ("
+      + slideDuration + ")")
```
Incorrect indentation.

tdas commented:
Aah, this has already been merged before I got to it. That's okay, barring a few nits, it was good. Thanks @wesleymiao @srowen
